photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/SliceSampler.scala

```diff
@@ -0,0 +1,214 @@
+
+import breeze.linalg.DenseVector
+import breeze.numerics.log
+
+/**
+ * This class implements the slice sampling algorithm. Slice sampling is an Markov chain Monte Carlo algorithm for
+ * sampling from an arbitrary continuous distribution function.
+ *
+ * Consider the ASCII-normal distribution function shown below.
+ *
+ *
+ *                  /-\                /-\ f(x)           /-\                /-\                /-\
+ *                 /   \              /  |\              /  |\              /   \              /   \
+ *                /     \            /   | \            /---|-\            /-x'--\            /     \
+ *             --/    x  \--      --/    x  \--      --/    x  \--      --/       \--      --/  x'   \--
+ *
+ *                   1                  2                  3                  4                  5
+ *
+ * The procedure for drawing a sample from the function is as follows:
+ *
+ *   1) Start with an initial point x in the domain (e.g. the sample from a previous iteration).
+ *   2) Sample a vertical value uniformly between zero and f(x).
+ *   3) Step out horizontally in both directions from the value obtained in step 2 until reaching the boundary of the
+          function. This is the "slice".
+ *   4) Sample uniformly from the slice to obtain the new point x'.
+ *   5) Back to step 1 with the new point.
+ *
+ * @param logp the log-transformed function from which to sample. Note that this need not sum to 1, and need only be
+ *   proportional to the target PDF.
+ * @param range the range of values within which to sample
+ * @param stepSize the size to expand a slice while taking one step in a single direction from the
+ *   starting point
+ * @param maxStepOuts the maximum number of steps allowed while expanding a slice
+ * @param seed the random seed
+ */
+class SliceSampler(
+    logp: (DenseVector[Double]) => Double,
+    range: (Double, Double) = (log(1e-5), log(1e5)),
+    stepSize: Double = 1.0,
+    seed: Long = System.currentTimeMillis) {
+
+  val random = new Random(seed)
+
+  /**
+   * Draws a new point from the target distribution
+   *
+   * @param x the original point
+   * @param direction the direction along which to draw a new point
+   * @return the new point
+   */
+  def draw(x: DenseVector[Double]): DenseVector[Double] = {
+    val directions = random.shuffle((0 until x.length).toList)
+    directions.foldLeft(x) { (currX, i) =>
+      draw(currX, getDirection(i, x.length))
+    }
+  }
+
+  /**
+   * Draws a new point from the target distribution along the given direction
+   *
+   * @param x the original point
+   * @param direction the direction along which to draw a new point
+   * @return the new point
+   */
+  protected def draw(
+      x: DenseVector[Double],
+      direction: DenseVector[Double]): DenseVector[Double] = {
+
+    require(checkDirection(direction), "Direction must be a standard unit basis vector.")
+
+    val y = log(random.nextDouble()) + logp(x)
```

Comment:
Yes, I see now - I misunderstood the slicer. Looking at a bi-modal example cleared it up for me.